s30020-NikiforKoseda/zad2.txt

```diff
@@ -0,0 +1,226 @@
+Podczas tworzenia natrafiłem na błąd, którego nie potrafię wyjaśnić. Jeżeli echo znajdowało się poza ifem to drugie męskie imie się nie drukowało. Nie byłem w stanie zidentyfikować problemu. Czy zmienne w ifach są dostępne poza nimi?
+###
+
+Ćwiczenie 5b
+Napisz skrypt, który wczyta jako argument jakiś tekst. Skrypt ten wypisze kolejno:
+
+Pierwszy znak z argumentu
+
+Ostatni znak z argumentu
+
+Zamieni w argumencie każdy tekst SOP na tekst poniżej (zastosuj echo z przełącznikiem -e):
+
+\e[32mSOP\e[0m
+
+Przykładowy tekst testowy:
+SOP! SOP w imieniu prawa! SOP w imieniu prawa!
+
+Wynikowo ma to wyglądać tak:
+
+SOP! SOP w imieniu prawa! SOP w imieniu prawa!
+
+###
+#!/bin/bash
+color_sop() {
+    echo -e "${1//SOP/\e[32mSOP\e[0m}"
+}
+color_sop "$1"
+
+Nie jestem pewien co robię nie tak. Próbuję wywołać na różne sposoby, samo echo wywołuje teskt na zielono, ale skryptem wywala po prostu literki używane do zastąpienia tekstu na zielono
+
+###
+
+Wyrażenia regularne (regex/regexp - regular expressions)
+Bash jako taki obsługuje w niektórych aspektach wyrażenia regularne, ale niestety nie jest to tak wygodne, jakby się czasami marzyło. Na szczęście jest program sed.
+
+Program sed jest określany przez twórców jako edytor tekstowy bez interfejsu użytkownika. Coś w tym jest. My nie będziemy się zgłębiać w całą składnie polecenia sed, natomiast skupimy się na małym wycinku - przetwarzaniu wyrażeń regularnych.
+
+
+Sed najczęściej przetwarza ciąg znaków który dociera do niego przez standardowe wejście i zwraca wynik na standardowe wyjście. Składnia wyrażeń regularnych jest w większości zgodna z innymi programami. Podstawowe elementy składni wyrażeń wyglądają tak:
+
+\[…\] - w miejsce … wchodzi lista znaków które mogą wystąpić na danym miejscu
+
+() - zbiór opcji
+
+^ - początek ciągu znaków (jeśli nie występuje wewnątrz nawiasów kwadratowych)
+
+[^] - wewnątrz nawiasów kwadratowych to negacja tego co w nawiasie
+
+$ - koniec ciągu znaków
+
+* - wzorzec tuż przed tym znakiem ma się powtarzać dowolną liczbę razy (od 0 do nieskończoności)
+
++ - wzorzec tuż przed tym znakiem ma się powtarzać co najmniej raz
+
+. - dowolny jeden znak
+
+…-… - zakres znaków np. [a-z]
+
+\ - znak modyfikacji (escape character)
+
+
+echo Alicja | sed s/a$/ę/g | sed 's/\([^ę]\)$/\1a/g'
+
+
+Narzędzia, tutoriale i materiały
+https://regex101.com/ - webowy edytor/tester
+
+https://regexr.com/ - kolejny webowy edytor/tester
+
+https://www.regular-expressions.info/ - opis składni, możliwości, dokumentacja
+
+
+Drobna informacja na przyszłe zajęcia:
+W bash-u można także dopasowywać wyrażenia regularne w samej składni
+
+[[ tekst_sprawdzany =~ wyrazenie_regularne ]] && echo pasuje
+
+
+Ćwiczenie 6
+Teraz postaraj się zrobić to zadanie z haczykiem i podmianą fragmentu zmiennej (Ala ma ...)
+
+###
+Wydaje mi się że rozwiązanie w 5a spełnia założenia. Nie jak tutaj zastosować REGEX, ale mogę się mylić, i jest optymalniejsza ścieżka zrobienia tego
```

Comment:
Optimum - najlepsza sytuacja.
Optymalniejsze - błąd językowy. Albo coś jest optymalne albo nie. To się nie stopniuje.

Merytorycznie:
Tak. Używając negatywnych indeksów skrypt powinien wykonywać się szybciej.